subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2273 @@
+ * @return Negative error code on failure.
+ */
+static int uvc_flush_vbuf(const struct device *dev, struct video_buffer *vbuf)
+{
+	const struct uvc_config *cfg = dev->config;
+	struct uvc_data *data = dev->data;
+	size_t next_vbuf_offset = data->vbuf_offset;
+	size_t next_line_offset = vbuf->line_offset;
+	struct net_buf *buf;
+	struct uvc_buf_info *bi;
+	int ret;
+
+	if (atomic_test_bit(&data->state, UVC_STATE_STREAM_RESTART)) {
+		return uvc_reset_transfer(dev);
+	}
+
+	buf = (data->vbuf_offset == 0)
+		? uvc_initiate_transfer(dev, vbuf, &next_line_offset, &next_vbuf_offset)
+		: uvc_continue_transfer(dev, vbuf, &next_line_offset, &next_vbuf_offset);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	bi = (struct uvc_buf_info *)udc_get_buf_info(buf);
+	bi->udc.ep = uvc_get_bulk_in(dev);
+
+	LOG_DBG("Video buffer %p, offset %u/%u, size %u",
+		vbuf, data->vbuf_offset, vbuf->bytesused, buf->len);
+
+	/* End-of-Transfer condition */
+	if (next_vbuf_offset >= vbuf->bytesused) {
+		bi->vbuf = vbuf;
+		bi->udc.zlp = (buf->len == uvc_get_bulk_mps(cfg->c_data));
+	}
+
+	ret = usbd_ep_enqueue(cfg->c_data, buf);
+	if (ret != 0) {
+		net_buf_unref(buf);
+		return ret;
+	}
+
+	/* End-of-Transfer condition */
+	if (next_vbuf_offset >= vbuf->bytesused) {
+		data->vbuf_offset = 0;
+		return UVC_VBUF_DONE;
+	}
+
+	data->vbuf_offset = next_vbuf_offset;
+	vbuf->line_offset = next_line_offset;
+	return 0;
+}
+
+static void uvc_flush_queue(const struct device *dev)
+{
+	struct uvc_data *data = dev->data;
+	struct video_buffer *vbuf;
+	int ret;
+
+	__ASSERT_NO_MSG(atomic_test_bit(&data->state, UVC_STATE_INITIALIZED));
+	__ASSERT_NO_MSG(!k_is_in_isr());
+
+	if (!atomic_test_bit(&data->state, UVC_STATE_ENABLED) ||
+	    !atomic_test_bit(&data->state, UVC_STATE_STREAM_READY)) {
+		LOG_DBG("UVC not ready yet");
+		return;
+	}
+
+	/* Lock the access to the FIFO to make sure to only process one buffer at a time.
+	 * K_FOREVER is not expected to take long, as uvc_flush_vbuf() never blocks.
+	 */
+	LOG_DBG("Locking the UVC stream");
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	while ((vbuf = k_fifo_peek_head(&data->fifo_in)) != NULL) {
+		/* Pausing the UVC driver will accumulate buffers in the input queue */
+		if (atomic_test_bit(&data->state, UVC_STATE_PAUSED)) {
+			break;
+		}
+
+		ret = uvc_flush_vbuf(dev, vbuf);
+		if (ret < 0) {
+			LOG_DBG("Could not transfer video buffer %p for now", vbuf);
+			break;
+		}
+		if (ret == UVC_VBUF_DONE) {
+			LOG_DBG("Video buffer %p transferred, removing from the queue", vbuf);
+			k_fifo_get(&data->fifo_in, K_NO_WAIT);
+		}
+	}
+
+	/* Now the other contexts calling this function can access the fifo safely. */
+	LOG_DBG("Unlocking the UVC stream");
+	k_mutex_unlock(&data->mutex);
+}
+
+static void uvc_enable(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct uvc_data *data = dev->data;
+
+	__ASSERT_NO_MSG(atomic_test_bit(&data->state, UVC_STATE_INITIALIZED));
```

Comment:
That should not be necessary, not initialized instance will not be part of any active configuration and will not be enabled.

Reply:
Applied [here](https://github.com/zephyrproject-rtos/zephyr/compare/fe0d99a132f333285b92799081cc9e5f531bbae5..54253b4ddc998b3378a59260889a57525b91ad5b).